[Team-13][iOS][Dale&Mase] 실제 API 데이터로 뷰 그리기 & SearchBar의 텍스트기반 자동완성 기능 구현

sungju-kim opened the pull request:
안녕하세요 만사 🙌 데일과 메이스 입니다
저번 주 PR이 엊그제 같은데 벌써 또 PR의 시간이 왔네요 

이번 PR에는 저번에 Mock으로만 출력했던 데이터들을
실제 API에서 받아오는 값으로 출력하도록 변경했는데
이 과정에서 리팩토링에 시간을 많이 쏟았던 것 같습니다.

더불어 MapKit을 이용해 textField에 지역 이름을 입력하면
실시간으로 text를 감지해서 목록을 받아오는 기능까지 완성했습니다.

모델과 뷰의 작업이 나뉜느낌이 있는데
대부분의 작업은 페어프로그래밍 방식으로 진행했으며 한명이라도 이해가 되지않는 코드는 사용하지 않는 방법을 선택했습니다.
이번 주도 잘 부탁드리겠습니다😄

~~P.S. 주말동안 여러가지 시도를 하다보니 커밋양이 많아졌습니다..~~ 항상 감사하고 사랑합니다..

<br>

# 구현 사항

## feature

- NetworkManager
  - Entity, EndPoint
  - Repository
  - ImageManager
  - NetworkInject propertyWrapper 활용
- SearchView
  - SearchBar 검색결과 바탕으로 SearchView 구현
  <img src="https://user-images.githubusercontent.com/78553659/171122346-95549c9e-ec8f-4855-8ba8-c3be3b7058e1.gif" width="250" height="550"/>



## refactor

### CustomSearchBar -> UISearchController로 변경
### SearchHomeViewModel protocol -> ViewModel protocol 로 변경
**기존 코드**
```swift
protocol SearchHomeViewModelAction {
    var loadHeader: PublishRelay<Void> { get }
    var loadNearCities: PublishRelay<Void> { get }
    var loadHeroBanner: PublishRelay<Void> { get }
    var loadTheme: PublishRelay<Void> { get }
}


protocol SearchHomeViewModelState {
    var loadedHeader: PublishRelay<[String]> { get }
    var loadedNearCities: PublishRelay<[SearchHomeEntity.City]> { get }
    var loadedHeroBanner: PublishRelay<[SearchHomeEntity.Banner]> { get }
    var loadedTheme: PublishRelay<[String]> { get }
}


protocol SearchHomeViewModelBindable: SearchHomeViewModelAction, SearchHomeViewModelState {
    func action() -> SearchHomeViewModelAction
    func state() -> SearchHomeViewModelState
}

```

**변경된코드**
```swift
protocol ViewModelAction {
    associatedtype actionType
    var loadAction: PublishRelay<actionType> { get }
}


protocol ViewModelState {
    associatedtype stateType
    var loadedState: PublishRelay<stateType> { get }
}


protocol ViewModelBindable: ViewModelAction, ViewModelState {
    func accept(_ value: actionType)
    func bind(_ completion: @escaping (stateType) -> Void)
}

```
SearchHomeViewModel의 각 뷰별로 뷰모델을 분리하고, 뷰모델에 ViewModelBindable을 채택하여 여러가지 뷰모델에서 활용할 수 있도록 변경하였습니다.

<br>

### SearchHome HeaderView 출력 로직 변경 ( local 에서 정해서 )
**기존코드**
```swift
final class SearchHomeCollectionViewDataSource: NSObject, UICollectionViewDataSource {
    var mockHeader: [String] = []
    
    ...
    
    headerView.setHeaderLabel(text: mockHeader[indexPath.section])
}
```

**변경코드**
```swift
enum SearchHomeCollectionViewSection: Int, CaseIterable {
    case heroBanner, nearCity, themeJourney
    
    var header: String {
        switch self {
        case .heroBanner:
            return ""
        case .nearCity:
            return "가까운 여행지 둘러보기"
        case .themeJourney:
            return "어디에서나, 여행은 살아보는거야!"
        }
    }
}



final class SearchHomeCollectionViewDataSource: NSObject, UICollectionViewDataSource {
    
    ...

    headerView.setHeaderLabel(text: SearchHomeCollectionViewSection.allCases[indexPath.section].header)
}
```

### searchCollectionViewDataSource와 searchCollectionViewDelegate 분리

PublishRelay를 통해 VC와 Delegate & DataSource가 이벤트를 주고받는 흐름입니다
(화살표는 의존성이 아니라 단순히 실행되는 흐름의 방향을 나타냅니다)
![](https://i.imgur.com/145aGs1.png)

1. delegate가 선택된 셀의 인덱스를 searchViewController에게 넘겨줌 
```swift
final class SearchCollectionViewDelegate: NSObject, UICollectionViewDelegate {
    
    let selectedCell = PublishRelay<Int>()
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        collectionView.deselectItem(at: indexPath, animated: true)
        selectedCell.accept(indexPath.item)
    }
    
}

```

<br>

2. searchViewController가 searchCollectionViewDataSource에게 해당 Index를 넘겨줌
```swift
final class SearchViewController: UIViewController, MKLocalSearchCompleterDelegate {
    
    ...
    
    func bind() {
        ...
        searchCollectionViewDelegate.selectedCell
        .bind { [weak self] index in 
               self?.searchCollectionViewDataSource.selectedCellIndex.accept(index)
        }

    }
}

```
<br>

3. searchCollectionViewDataSource가 선택된 인덱스의 MKSearchCompletion 값을 searchViewController에게 전달

```swift
final class SearchCollectionViewDataSource: NSObject, UICollectionViewDataSource {
    
    var locations: [MKLocalSearchCompletion] = []
    
    let selectedCellIndex = PublishRelay<Int>()
    let selectedLocation = PublishRelay<MKLocalSearchCompletion>()
    
    override init() {
        super.init()
        selectedCellIndex.bind { [weak self] index in
            guard let index = self?.locations[index] else { return }
            self?.selectedLocation.accept(index)
        }
    }
    
    ...
    

```
<br>
    
4. dataSource로부터 받아온 MKSearchCompletion 값을 FilteringViewController에게 전달
```swift
final class SearchViewController: UIViewController, MKLocalSearchCompleterDelegate {
    
    ...
    
    func bind() {
    ...
        searchCollectionViewDataSource.selectedLocation
        .bind { [weak self] location in
                self?.navigationItem.backButtonTitle = "뒤로"
                self?.navigationController?.pushViewController(FilteringViewController(location: location), animated: true)
        }
    }
}

```




### SearchHomeCollectionView 전체가 아닌 특정 Section만 reload하도록 변경
 **기존코드**
 ```swift
self?.collectionView.reloadData()
 ```
 
 **변경코드**
 ```swift
enum SearchHomeCollectionViewSection: Int {
    case heroBanner, nearCity, themeJourney
    
    var indexSet: IndexSet {
        return IndexSet(integer: rawValue)
}
```
```swift
self?.collectionView.reloadSections(SearchHomeCollectionViewSection.heroBanner.indexSet)
  ```


<br>
<br>

# 구현할 사항
- 지난 PR에서 `case .themeJourney: return "어디에서나, 여행은 살아보는거야!"`와 같은 String 상수값을 별도로 관리하는 타입을 만들겠다고 했는데, 이번 PR에서도 구현하지 못했습니다. 다음 PR에는 꼭! 구현해서 리팩토링하겠습니다😭

- 이제 `FilteringViewController`를 본격적으로 구현해 볼 예정입니다.
  ![](https://i.imgur.com/OSeZO3c.jpg)



<br>
<br>


# 조언을 얻고 싶은 부분


아래 코드와 같이 **프로퍼티 초기화를 위해 생성자에 별도의 매개변수가 추가될 경우**,
`required init?`을 `init`과 동일 동작하도록 맞춰주기 위해선
`required init?`에서도 viewModel을 매개변수로 받아야 합니다.

```swift
init(viewModel: SearchHomeViewModel) {
    self.viewModel = viewModel
    super.init(nibName: nil, bundle: nil)
}

@available(*, unavailable)
required init?(coder: NSCoder) {
    super.init(coder: coder)
    fatalError("init(coder:) has not been implemented")
}
```
하지만 `required init?`은 `init(coder:)` 형태로만 사용해야하므로
매개변수를 추가하는 등의 변경을 할 수 없습니다.

이 경우 **`required init?`도 `init`과 동일하게 동작해야 한다**는 원칙을 유지할 수가 없어 보이는데, 이럴 땐 어떻게 해야 할까요?

그냥 지금처럼 available 어노테이션을 달아서 컴파일 타임에 사용되지 못하도록 막는 것이 최선일까요?

sungju-kim committed:
feat: SearchView구현

- SearchViewController 구현
- PopularHeaderVIew 구현
- PopularCollectionViewDataSource 구현

- SectionLayoutFactory / createPopularDestinationLayout메소드 추가

jeremy0405/airbnb/#5

sungju-kim committed:
feat: SeachBar 검색결과 바탕으로 SearchResultCollectionView reload

- MapKit 활용
- SearchCollectionViewDataSource
- SearchBarDelegate
- SearchResultViewCell

jeremy0405/airbnb/#5

sanghyeok-kim committed:
feat: API 통신을 위한 Entity 객체와 EndPoint Case 설계

- SearchHomeEntity
  - NearCity
  - HeroBanner

jeremy0405/airbnb/#9

sanghyeok-kim committed:
refactor: NetworkError case 네이밍 수정

- decode -> failToDecode
- invalidUrl 추가
- 그 외 case들은 불필요해서 삭제

jeremy0405/airbnb/#9

sanghyeok-kim committed:
feat: NetworkManager 설계

- Alamofire 활용한 request 메소드 구현

jeremy0405/airbnb/#9

sanghyeok-kim committed:
feat: Repository 구현

- 모든 Repository들이 부모로 가질 상위 클래스 NetworkRepository (decdoe 메소드를 포함)
- SearchHome에서 사용할 Repository 프로토콜 (SearchHomeRepository)
- SearchHomeRepository을 채택해서 구현한 SearchHomeRepositoryImpl

jeremy0405/airbnb/#9

sanghyeok-kim committed:
feat: ImageManager 구현

- URL을 받아 download task로 response 요청 후 UIImage 형태로 completion의 매개변수에 넣어줌
- Disk Caching 기능 구현

jeremy0405/airbnb/#9

sanghyeok-kim committed:
refactor: Mock Data가 아닌 실제 NetworkManager로부터 받아온 Data를 화면에 출력하도록 변경

- HeroBanner, NearCity 값을 API에서 받아온 data를 통해 출력하도록 변경함
- Theme도 마찬가지로 받아오도록 변경할 예정(현재 mock data 사용중)

jeremy0405/airbnb/#9

sanghyeok-kim committed:
chore: 불필요한 코드 삭제

- VMAction, VMState에 선언된 accept, bind 메소드 삭제
  (해당 로직은 VM이 구현해도 되고 안해도 되므로)
- ImageManger 로직에 들어간 print문 삭제

jeremy0405/airbnb/#9

sanghyeok-kim committed:
[Merge pull request] dev-iOS <- iOS-feature/Alamofire

[iOS] NetworkManager 설계 및 API로부터 받아오는 로직 추가

sungju-kim committed:
Merge branch 'dev-iOS' into iOS-feature/SearchVC

sungju-kim committed:
[Merge pull request] dev-iOS <- iOS-feature/SearchVC

[iOS] SearchView 및 SearchBar 자동완성

sungju-kim committed:
refactor: SearchHomeViewModel protocol 재사용가능하게 변경

- SearchViewModelAction -> ViewModelAction 변경
- SearchViewModelState -> ViewModelState 변경
- SearchViewModelBindable -> ViewModelBindable 변경

- HeroBannerViewModel 추가
- HeaderViewModel 추가
- NearCitiesViewModel 추가
- ThemeViewModel 추가

jeremy0405/airbnb/#5

sungju-kim committed:
feat: themeJourney Network로 data 받아오기

- SeachHomeEndPoint case 추가
- ThemeViewModel loadAction.bind 구문 변경
- SearchHomeRepositoryImpl/ requestTheme() 추가

jeremy0405/airbnb/#9

sanghyeok-kim committed:
chore: 변수명 변경, 인덴트 수정, 불필요한 코드 제거

- 변수명 통일성 있게 변경
- 인덴트 수정

sanghyeok-kim committed:
refactor: SearchHome의 HeaderView를 직접 출력하도록 변경

- 서버에서 받아오지 않고 직접 상수로 갖는 값이므로 불필요해짐

sanghyeok-kim committed:
refactor: CollectionView 전체가 아닌 특정 Section만 reload하도록 변경

- reloadData -> reloadSection으로 변경

sungju-kim committed:
refactor: CustomSearchBar -> UISearchController 변경

- searchBar -> searchController 변경
- 관련 로직 변경

jeremy0405/airbnb/#5

sanghyeok-kim committed:
chore: result -> themeJourney 이름 변경

sanghyeok-kim committed:
chore: 불필요한 코드 삭제

- SwiftUI import 삭제

sanghyeok-kim committed:
feat: SearchView 도시 검색 기능 구현

- 근처 인기 여행지(nearCities)를 서버로부터 도시 정보, 이미지 받아와서 출력하도록 변경
- 도시 검색시 나타나던 뷰 수정
  - 도시 이미지를 Mock Image -> SF Symbol로 변경
  - cell 레이아웃 조정

jeremy0405/airbnb/#5

sungju-kim committed:
feat: searchCollectionViewDataSource와 searchCollectionViewDelegate 분리 및 선택된 인덱스 처리 흐름 구현

- searchCollectionViewDelegate가 선택된 Cell의 인덱스를 searchViewController에게 넘겨줌
- searchViewController가 searchCollectionViewDataSource에게 해당 Index를 넘겨줌
- searchCollectionViewDataSource가 선택된 인덱스의 MKSearchCompletion 값을 searchViewController에게 넘겨줌

jeremy0405/airbnb/#5

sungju-kim committed:
chore: 불필요한 코드 제거

sungju-kim committed:
feat: FilteringViewController 추가

- SearchViewController에서 터치된 cell의 location값을 받아오는 FilteringViewController

jeremy0405/airbnb/#5

sungju-kim committed:
feat: NetworkInject propertyWrapper 구현

- keyPath를 통해 NetworkContainter 싱글톤 객체의 프로퍼티에 접근 하도록 변경
- propertyWrapper로 keyPath를 전달후 이에 맞는 Network 객체 주입

sungju-kim committed:
refactor: DataSource의 data 접근 제어자 변경 및 메소드 구현

- SearchCollectionViewDataSource
- PopularCollectionViewDataSource
- SearchHomeCollectionViewDataSource

sanghyeok-kim committed:
chore: 불필요한 코드 제거, 변수명 수정

- print문 제거
- 불필요한 self 제거
- 변수명 통일

sanghyeok-kim committed:
chore: View의 required init을 override init과 동일하게 동작하도록 변경

- fatalError가 아닌 override init과 동일한 로직으로 구현

sanghyeok-kim commented on iOS/AirbnbApp/AirbnbApp/Source/Support/NetworkInject.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+@propertyWrapper
+struct NetworkInject<T> {
+    var wrappedValue: T
+    
+    init(keypath: KeyPath<NetworkContainer, T>) {
+        let container = NetworkContainer.shared
+        wrappedValue = container[keyPath: keypath]
+    }
+}
```
keyPath를 매개변수로 받아 NetworkContainer의 프로퍼티에 접근해
wrappedValue에 할당해주는 PropertyWrapper를 사용했습니다.

sanghyeok-kim committed:
fix: API 변경사항 적용 및 화면 출력 관련 버그 수정

- Mock API -> 실제 API로 부터 받아오도록 EndPoint 수정
- SearchView 출력 안되는 버그 수정
- Push될 때 하단 TabBar 숨겨지지 않는 버그 수정

sanghyeok-kim commented on iOS/AirbnbApp/AirbnbApp/Source/Model/Model/API/SearchHomeEndPoint.swift:
```diff
@@ -0,0 +1,57 @@
+extension SearchHomeEndPoint {
+    var baseUrl: URL? {
+        switch self {
+        case .nearDestination, .heroBanner, .themeJourney:
+            return URL(string: "http://3.34.196.158/api")
+        }
+    }
+    
```
서버분들께서 방금 전 Mock 서버가 아닌 실제 서버를 배포해주셔서,
BaseUrl이 변경된 사항을 후다닥 적용해서 Push 했습니다.
(PR 보낸 이후에 Push된 사항이라 말씀드립니다!)

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Model/Model/API/EndPoint/NetworkManager.swift:
```diff
@@ -0,0 +1,34 @@
+        request
+            .validate(statusCode: 200..<300)
+            .response { response in
+                if let data = response.data {
+                    completion(.success(data))
+                } else {
+                    completion(.failure(.emptyData))
+                }
+            }
```
request 메서드의 뎁스를 한번 더 줄이면 가독성이 좀 더 좋아질거 같습니다.!!

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Model/Model/API/EndPoint/Requestable.swift:
```diff
@@ -8,11 +8,10 @@
     var baseUrl: URL? { get }
     var path: String? { get }
-    var fullUrl: Alamofire.URLConvertible? { get }
+    var fullUrl: URL? { get }
```
fullUrl을 옵셔널로 사용하고 있는 이유를 알 수 있을까요?~~

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Model/Model/API/SearchHomeEndPoint.swift:
```diff
@@ -0,0 +1,57 @@
+import Alamofire
+
+enum SearchHomeEndPoint: Requestable {
+    case nearDestination(latitude: Double, longtitude: Double)
```
latitude, longtitude 를 따로 받는 것도 좋지만 coordinate같은 개체를 사용하여 원시값을 포장하는 것을 추천합니다!

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Present/Search/PopularCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,62 @@
+
+import UIKit
+
+final class PopularCollectionViewDataSource: NSObject, UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {
```
DataSource class에 Delegate / DelegateFlowLayout도 채택이 돼 있네요! 이 부분도 분리할 수 있으면 좀 더 좋을거 같아요 ㅎㅎ

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Present/Search/PopularCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,62 @@
+        }
+        let item = nearCities[indexPath.item]
+        
+        let imageUrl = URL(string: item.imageUrl)
+        imageManager.fetchImage(from: imageUrl) { image in
+            DispatchQueue.main.async {
+                cell.setCityImageView(image: image ?? UIImage())
+                // TODO: image가 nil일 경우 handling하는 에러 구현하기
+            }
+        }
+        cell.setCityTitleLabel(text: item.cityName)
+        cell.setDistanceLabel(text: item.time.convertIntoTime())
+        return cell
```
cell.setCityImageVIew등을 datasource가 아니라 configureWith와 같은 메서드를 cell에 만들고 nearCity 모델만 넘겨주는 방식이라면 좀 더 좋을거 같아요!!

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Present/Search/PopularCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,62 @@
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
+        if kind == UICollectionView.elementKindSectionHeader {
+            
+            guard let headerView = collectionView.dequeueReusableSupplementaryView(
+                ofKind: kind,
+                withReuseIdentifier: PopularHeaderView.identifier,
+                for: indexPath
+            ) as? PopularHeaderView else { return UICollectionReusableView() }
+            
+            headerView.setHeaderLabel(text: "근처의 인기 여행지")
+            return headerView
+        }
+        
+        return UICollectionReusableView()
+    }
```
귀찮은 작업이지만 이런 상황에서도 else와 같은 것을 줄이면서 뎁스를 줄여보는 연습을 해보면 좋을거 같아요~ (참고만 해주세요 ㅎㅎ)

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Present/Search/SearchBarDelegate.swift:
```diff
@@ -0,0 +1,50 @@
+import UIKit
+import MapKit
+
+class SearchBarDelegate: NSObject, UISearchBarDelegate {
```
SearchBarDelegate 코느는 가독성이 굉장히 좋네요. ㅎㅎㅎ

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Present/Search/SearchCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,47 @@
+import UIKit
+import MapKit
+
+final class SearchCollectionViewDataSource: NSObject, UICollectionViewDataSource {
```
SearchCollectionViewDataSource 역시 가독성이 굉장하네요 ㅎㅎㅎ 잘 읽힙니다.!

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Present/Search/SearchViewController.swift:
```diff
@@ -0,0 +1,148 @@
+
+final class SearchViewController: UIViewController {
+    
+    private let viewModel: NearCityViewModel
```
NearCityViewModel가 채택중인 ViewModelBindable이 아니라 NearCityViewModel 인스턴스를 사용하신 이유를 알 수 있을까요?~

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Present/SearchHome/SearchHomeCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,129 @@
+    private var banners: [SearchHomeEntity.Banner] = []
+    private var themeJourney: [SearchHomeEntity.Theme] = []
+    
+    @NetworkInject(keypath: \.imageManager)
+    private var imageManager: ImageManager
+    
+    @NetworkInject(keypath: \.searchHomeRepositoryImplement)
+    private var repository: SearchHomeRepositoryImpl
```
NetworkInject를 사용한 것은 좋네요 ㅎㅎ

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Present/SearchHome/SearchHomeCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,129 @@
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        guard let sectionKind = SearchHomeCollectionViewSection(rawValue: section) else { return 0 }
+        
+        switch sectionKind {
+        case .heroBanner:
+            return banners.count
+        case .nearCity:
+            return nearCities.count
+        case .themeJourney:
+            return themeJourney.count
+        }
```
sectionKind가 무럭무럭 자라서 10개, 20개가 되는 경우가 생긴다면 이 코드를 조금은 문제가 생기겠죠?ㅎㅎㅎ.. 이럴 떈 다형성 같은 것을 사용하여 switch case를 줄여보는 연습도 좋을거 같아요 ㅎㅎ

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Present/SearchHome/SearchHomeCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,129 @@
+            cell.setDescriptionLabel(text: item.description)
+            return cell
+            
+        case .nearCity:
+            guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: CityViewCell.identifier, for: indexPath) as? CityViewCell else {
+                return UICollectionViewCell()
+            }
+            
+            let item = nearCities[indexPath.item]
+            
+            let imageUrl = URL(string: item.imageUrl)
+            imageManager.fetchImage(from: imageUrl) { image in
+                DispatchQueue.main.async {
+                    cell.setCityImageView(image: image ?? UIImage())
+                    // TODO: image가 nil일 경우 handling하는 에러 구현하기
+                }
+            }
+            
+            cell.setCityTitleLabel(text: item.cityName)
+            cell.setDistanceLabel(text: item.time.convertIntoTime())
+            return cell
+            
+        case .themeJourney:
+            guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: ThemeJourneyViewCell.identifier, for: indexPath) as? ThemeJourneyViewCell else {
+                return UICollectionViewCell()
+            }
+            
+            let item = themeJourney[indexPath.item]
+            let imageUrl = URL(string: item.imageUrl)
+            imageManager.fetchImage(from: imageUrl) { image in
+                DispatchQueue.main.async {
+                    cell.setImageView(image: image ?? UIImage())
+                    // TODO: image가 nil일 경우 handling하는 에러 구현하기
+                }
+            }
+            
+            cell.setDescriptionLabel(text: item.title)
+            return cell
+        }
+    }
```
메서드 길이가 조금 길죠..?ㅎㅎㅎㅎ 중복되는 것은 하나로 합치고 switch case도 줄여본다면 분명 큰 연습이 될거 같아요 ㅎㅎ!!

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Present/SearchHome/ViewModel/HeroBannerViewModel.swift:
```diff
@@ -0,0 +1,46 @@
+    typealias actionType = Void
+    typealias stateType = [SearchHomeEntity.Banner]
+    
+    private(set) var loadAction = PublishRelay<actionType>()
+    private(set) var loadedState = PublishRelay<stateType>()
```
getter 보다는 메서드를 사용하는건 어떨까요?~~

ITzombietux commented on iOS/AirbnbApp/AirbnbApp/Source/Repository/Common/ImageManager.swift:
```diff
@@ -0,0 +1,46 @@
+    
+    private let fileManager = FileManager()
+    
+    func fetchImage(from url: URL?, completion: @escaping (UIImage?) -> Void) {
+        
+        guard let url = url else { return } // TODO: 에러 처리하기
+        
+        let fileName = url.lastPathComponent
+        
+        guard let cachesDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
+            return
+        }
+        
```
FileManager를 파라미터로 받고 fetchImage를 class func로 사용해도 좋을거 같다는 생각이 들었습니다 ㅎㅎ ImageManager 인스턴스를 만들지 않고도 사용할 수 있다면 더 좋겠죠?

ITzombietux reviewed (approved):
Dale&Mase 안녕하세요! PR보내시느라 너무나 수고많으셨습니다.!!  오늘도 역시 PR 메시지를 너무 디테일하게 잘 적어주셔서 코드를 보는데 있어 굉장히 큰 도움이 됐습니다. 감사합니다. 👍

페어 프로그래밍으로 진행하셨다고 했는데, 코드의 품질이 꽤나 높아 보였고, 특정 객체는 가독성이 아주 좋았습니다. 코드 전반에 메서드 길이, 뎁스 같은 부분을 의식적으로 줄여보는 연습을 해보시면서 코드를 수정 해보시면 가독성이 더더 좋아질거 같아요! (사실 이 부분은 저도 연습이 잘 안된 부분입니다..ㅠㅠ)

고민하고 게신 required init? 을 사용하지 않는다면 어노테이션을 사용하여 코드를 사용하지 않는 방법도 괜찮은 방법이라고 생각은 들지만, 최선의 방법이라는게 있을까?! 싶습니다 ㅎㅎ 어떻게 구현하냐에 따라 어떤 선택을 하는지가 결정되는데, 구현 방법에 맞는걸 선택하셨다고 봅니다.!

sanghyeok-kim committed:
refactor: longitude, latitude을 관리하는 Coordinate 타입 생성

- 좌표값을 원시타입이 아닌 Coordinate 타입으로 입력받도록 리팩토링

jeremy0405/airbnb/#42

sungju-kim committed:
refactor: CollectionView의 Section과 Cell에 대한 ViewModel 및 protocol구현

- 조건문 없이 다형성으로 처리되도록 리팩토링
- CitySectionViewModel, PopularSectionViewModel 구현
    - CityCellViewModel 구현
- BannerSectionViewModel 구현
    - BannerCellViewModel 구현
- ThemeSectionViewModel 구현
    - ThemeCellViewModel 구현
- 관련 로직 변경

jeremy0405/airbnb/#42

sungju-kim committed:
refactor: SearchHomeViewController 중복코드 제거

- associated type ->  [CellViewModelable]로 추상화 하여 ViewModelBindable 변경
- SearchHomeCollectionViewSection의 AllCases 이용

jeremy0405/airbnb/#42

sanghyeok-kim committed:
chore: 인덴트, 줄바꿈, 함수 선언 위치 등 수정

sanghyeok-kim committed:
fix: Cell이 Reuse되면서 중복되는 현상을 DisposeBag 기능 구현을 통해 해결

- DisposeBag 기능 구현
- ViewCell이 ViewModel에 bind한 것이 계속 append로 쌓이지 않도록 clearBinds()를 통해 비워줌

sanghyeok-kim commented:
<img width="1932" alt="CleanShot 2022-06-03 at 16 29 02@2x" src="https://user-images.githubusercontent.com/57667738/171808399-a24ca18f-0022-4a00-a3b3-55114c2fb052.png">

<img width="1533" alt="CleanShot 2022-06-03 at 16 31 11@2x" src="https://user-images.githubusercontent.com/57667738/171808658-06f1fc39-3f13-4a8d-b057-a76a6a5abcf9.png">
